File: backend/app/APIs/routers/generation.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Response, status

from app.APIs.dependencies import get_generation_service
from app.Services.generation_service import GenerationService
from app.Schemas.common import GenerationPresetKind
from app.Schemas.generation import (
  CreateGenerationPresetRequest,
  GenerateChapterBlueprintRequest,
  GenerateDocumentContinuationRequest,
  GenerateDraftRequest,
  GenerateQuickDraftRequest,
  GeneratedChapterBlueprint,
  GeneratedDraft,
  GenerationPreset,
  GenerationPresetLibrary,
  PolishDocumentSelectionRequest,
  PolishSelectionRequest,
  UpdateGenerationPresetRequest,
)

router = APIRouter()


@router.get("/presets", response_model=GenerationPresetLibrary)
async def list_generation_presets(
  project_id: str,
  service: GenerationService = Depends(get_generation_service),
) -> GenerationPresetLibrary:
  return await service.list_presets(project_id)


@router.post(
  "/presets",
  response_model=GenerationPreset,
  status_code=status.HTTP_201_CREATED,
)
async def create_generation_preset(
  project_id: str,
  request: CreateGenerationPresetRequest,
  service: GenerationService = Depends(get_generation_service),
) -> GenerationPreset:
  return await service.create_preset(project_id, request)


@router.put("/presets/{kind}/{preset_id}", response_model=GenerationPreset)
async def update_generation_preset(
  project_id: str,
  kind: GenerationPresetKind,
  preset_id: str,
  request: UpdateGenerationPresetRequest,
  service: GenerationService = Depends(get_generation_service),
) -> GenerationPreset:
  return await service.update_preset(project_id, kind, preset_id, request)


@router.delete("/presets/{kind}/{preset_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_generation_preset(
  project_id: str,
  kind: GenerationPresetKind,
  preset_id: str,
  service: GenerationService = Depends(get_generation_service),
) -> Response:
  await service.delete_preset(project_id, kind, preset_id)
  return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/draft", response_model=GeneratedDraft)
async def generate_draft(
  project_id: str,
  request: GenerateDraftRequest,
  service: GenerationService = Depends(get_generation_service),
) -> GeneratedDraft:
  return await service.generate_draft(project_id, request)


@router.post("/quick-draft", response_model=GeneratedDraft)
async def generate_quick_draft(
  project_id: str,
  request: GenerateQuickDraftRequest,
  service: GenerationService = Depends(get_generation_service),
) -> GeneratedDraft:
  return await service.generate_quick_draft(project_id, request)


@router.post("/chapter-blueprint", response_model=GeneratedChapterBlueprint)
async def generate_chapter_blueprint(
  project_id: str,
  request: GenerateChapterBlueprintRequest,
  service: GenerationService = Depends(get_generation_service),
) -> GeneratedChapterBlueprint:
  return await service.generate_chapter_blueprint(project_id, request)


@router.post("/polish", response_model=GeneratedDraft)
async def polish_selection(
  project_id: str,
  request: PolishSelectionRequest,
  service: GenerationService = Depends(get_generation_service),
) -> GeneratedDraft:
  return await service.polish_selection(project_id, request)


@router.post("/documents/polish", response_model=GeneratedDraft)
async def polish_document_selection(
  project_id: str,
  request: PolishDocumentSelectionRequest,
  service: GenerationService = Depends(get_generation_service),
) -> GeneratedDraft:
  return await service.polish_document_selection(project_id, request)


@router.post("/documents/continue", response_model=GeneratedDraft)
async def generate_document_continuation(
  project_id: str,
  request: GenerateDocumentContinuationRequest,
  service: GenerationService = Depends(get_generation_service),
) -> GeneratedDraft:
  return await service.generate_document_continuation(project_id, request)

File: frontend/lib/api/generation.ts
import { apiFetch } from "./client";
import type {
  GeneratedChapterBlueprint,
  GeneratedDraft,
  GenerateChapterBlueprintInput,
  GenerateDocumentContinuationInput,
  GenerateDraftInput,
  GenerateQuickDraftInput,
  GenerationPreset,
  GenerationPresetInput,
  GenerationPresetKind,
  GenerationPresetLibrary,
  GenerationPresetUpdate,
  PolishDocumentSelectionInput,
  PolishSelectionInput
} from "./types/index";

export async function getGenerationPresets(
  projectId: string,
): Promise<GenerationPresetLibrary> {
  return await apiFetch<GenerationPresetLibrary>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/presets`,
  );
}

export async function createGenerationPreset(
  projectId: string,
  input: GenerationPresetInput,
): Promise<GenerationPreset> {
  return await apiFetch<GenerationPreset>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/presets`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
}

export async function updateGenerationPreset(
  projectId: string,
  kind: GenerationPresetKind,
  presetId: string,
  input: GenerationPresetUpdate,
): Promise<GenerationPreset> {
  return await apiFetch<GenerationPreset>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/presets/${encodeURIComponent(kind)}/${encodeURIComponent(presetId)}`,
    {
      method: "PUT",
      body: JSON.stringify(input),
    },
  );
}

export async function deleteGenerationPreset(
  projectId: string,
  kind: GenerationPresetKind,
  presetId: string,
): Promise<void> {
  await apiFetch<void>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/presets/${encodeURIComponent(kind)}/${encodeURIComponent(presetId)}`,
    {
      method: "DELETE",
    },
  );
}

export async function generateDraft(
  projectId: string,
  input: GenerateDraftInput,
): Promise<GeneratedDraft> {
  return await apiFetch<GeneratedDraft>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/draft`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
}

export async function generateChapterBlueprint(
  projectId: string,
  input: GenerateChapterBlueprintInput,
): Promise<GeneratedChapterBlueprint> {
  return await apiFetch<GeneratedChapterBlueprint>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/chapter-blueprint`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
}

export async function generateQuickDraft(
  projectId: string,
  input: GenerateQuickDraftInput,
): Promise<GeneratedDraft> {
  return await apiFetch<GeneratedDraft>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/quick-draft`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
}

export async function polishSelection(
  projectId: string,
  input: PolishSelectionInput,
): Promise<GeneratedDraft> {
  return await apiFetch<GeneratedDraft>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/polish`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
}

export async function polishDocumentSelection(
  projectId: string,
  input: PolishDocumentSelectionInput,
): Promise<GeneratedDraft> {
  return await apiFetch<GeneratedDraft>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/documents/polish`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
}

export async function generateDocumentContinuation(
  projectId: string,
  input: GenerateDocumentContinuationInput,
): Promise<GeneratedDraft> {
  return await apiFetch<GeneratedDraft>(
    `/api/projects/${encodeURIComponent(projectId)}/generation/documents/continue`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
}
